core/app/models/spree/promotion_code/batch_builder.rb

```diff
@@ -0,0 +1,64 @@
+class ::Spree::PromotionCode::BatchBuilder
+  attr_reader :promotion_code_batch
+  delegate :promotion, :number_of_codes, :base_code, to: :promotion_code_batch
+
+  class_attribute :random_code_length, :batch_size
+  self.random_code_length = 8
+  self.batch_size = 1_000
+
+  def initialize(promotion_code_batch)
+    @promotion_code_batch = promotion_code_batch
+  end
+
+  def build_promotion_codes
+    generate_random_codes
+
+    PromotionCodeBatchMailer.promotion_code_batch_finished(promotion_code_batch).deliver_now
+  rescue => e
+    promotion_code_batch.update!(error: e.inspect)
+    PromotionCodeBatchMailer.promotion_code_batch_errored(promotion_code_batch).deliver_now
+    raise e
+  end
+
+  private
+
+  def generate_random_codes
+    total_batches = (number_of_codes.to_f / self.class.batch_size).ceil
+
+    total_batches.times do |i|
+      codes_for_current_batch = Set.new
+      codes_to_generate = [self.class.batch_size, number_of_codes - i * batch_size].min
+
+      while codes_for_current_batch.size < codes_to_generate
+        new_codes = Array.new(codes_to_generate) { generate_random_code }.to_set
+        codes_for_current_batch += get_unique_codes(new_codes)
+      end
+
+      codes_for_current_batch.map do |code|
+        promotion.codes.build(value: code)
```

Comment:
Should we perhaps use `Spree::PromotionCode.create!` here instead of `promotion.codes.build ... promotion.save!`?  That would avoid loading a million objects into memory if you're generating a million codes.